File: leetcode/editor/en/[20]Valid Parentheses.java
20
        Valid Parentheses
        2022-11-29 18:24:11

//leetcode submit region begin(Prohibit modification and deletion)
class Solution {
    public boolean isValid(String s) {
        Deque<Character> dq = new ArrayDeque<>();
        for (char c : s.toCharArray()) {
            if (c == '(' || c == '[' || c == '{') {
                dq.push(c);
            } else {
                if (dq.isEmpty()) {
                    return false;
                }
                char ch = dq.pop();
                if ((ch == '(' && c == ')') || (ch == '[' && c == ']') || (ch == '{' && c == '}')) {
                    continue;
                } else {
                    return false;
                }
            }
        }
        return dq.isEmpty();
    }
}
//leetcode submit region end(Prohibit modification and deletion)

File: leetcode/editor/en/[20]Valid Parentheses.js
//leetcode submit region begin(Prohibit modification and deletion)
/**
 * @param {string} s
 * @return {boolean}
 */
var isValid = function (s) {
    const stack = [];
    for (let i = 0; i < s.length; i++) {
        const c = s[i];
        if (c === '(' || c === '[' || c === '{') {
            stack.push(c);
        } else {
            if (stack.length === 0) {
                return false;
            } else {
                const top = stack.pop();
                if ((top === '(' && c !== ')') ||
                    (top === '[' && c !== ']') ||
                    (top === '{' && c !== '}')) {
                    return false;
                }
            }
        }
    }
    return stack.length === 0;
};
//leetcode submit region end(Prohibit modification and deletion)
